### PR TITLE
codeintel: Minimize tuple modifications when updating commit graph data

### LINKED ISSUE #18285: codeintel: Research HOT for bulk inserts
See if we can do better than deleting/re-inserting in bulk for the `lsif_nearest_uploads`/`_links` tables. Using some form of update/not touching rows/heap-only-tuples could reduce the bloat on these tables.

### PR DESCRIPTION
**Pull request by efritz**
Fixes #18285.

When updating the `lsif_nearest_uploads`, `lsif_nearest_uploads_links`, and `lsif_uploads_visible_from_tip` tables, we do a mass delete for all data related to a repository, then re-insert the newly calculated data for that repository. We've compressed it fairly well, but that's no match for _outpacing Postgres's autovacuum daemon_.

Turns out that MVCC is really not a space-efficient when your workload is basically telling Postgres to drop 90% of the data you're about to write to it a few milliseconds later.

Since the commit graph is generally stable (anything outside of the influence of an active branch does not change), we're now only going to touch the rows in the table that we absolutely need to. We still insert the entire data set into Postgres, just now into a temporary table. We then use that temporary table and the current data to insert rows for which there isn't an entry in the index, delete records for which there's no correlated or win the temp table, and update the remaining rows (that have changed). This alone should account for a massive reduction in how much we write to the WAL on these fairly frequent operations.

### CODE REVIEW

File: enterprise/internal/codeintel/stores/dbstore/commits.go [hunk 1]
```diff
@@ -212,37 +212,108 @@ func (s *Store) CalculateVisibleUploads(ctx context.Context, repositoryID int, c
...
+	if err := tx.persistNearestUploadsLinks(ctx, repositoryID); err != nil {
+		return err
+	}
+
+	// Persist data to permenant table: t_lsif_uploads_visible_at_tip -> lsif_uploads_visible_at_tip
+	if err := tx.persistUploadsVisibleAtTip(ctx, repositoryID); err != nil {
+		return err
+	}
+
+	if dirtyToken != 0 {
```
Comment (slimsag) [inline enterprise/internal/codeintel/stores/dbstore/commits.go:235]: Seems like a code smell

File: enterprise/internal/codeintel/stores/dbstore/commits.go [hunk 2]
```diff
@@ -212,37 +212,108 @@ func (s *Store) CalculateVisibleUploads(ctx context.Context, repositoryID int, c
...
+UPDATE lsif_dirty_repositories SET update_token = GREATEST(update_token, %s), updated_at = %s WHERE repository_id = %s
+`
+
+// writeVisibleUploads serializes the given commit graph into a the following set of temporary tables in the database.
+//
+//   - t_lsif_nearest_uploads        (mirroring lsif_nearest_uploads)
+//   - t_lsif_nearest_uploads_links  (mirroring lsif_nearest_uploads_links)
+//   - t_lsif_uploads_visible_at_tip (mirroring lsif_uploads_visible_at_tip)
+//
+// The data in these temporary tables can then be moved into a persisted/permenant table. We previously would perform a
```
Comment (slimsag) [inline enterprise/internal/codeintel/stores/dbstore/commits.go:264]:
```suggestion
// The data in these temporary tables can then be moved into a persisted/permanent table. We previously would perform a
```

**Review by slimsag — APPROVED** [review]

File: enterprise/internal/codeintel/stores/dbstore/commits.go (hunk 1 above)
Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/commits.go:235]: I think this is actually left over (or from old tests), will see if I can do this unconditionally.

Comment (slimsag) [reply enterprise/internal/codeintel/stores/dbstore/commits.go:235]: Haha, I'm not sure what you're talking about but my comment was a joke :)

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/commits.go:235]: 🤦

**Review by Strum355 — APPROVED** [review]
LGTM :shipit: I notice an absence of your usual comprehensive tests though :eyes: Is this handled elsewhere/forgotten/purposely excluded?

**Pull request merged**